kempner_hpc_handbook/scalability/distributed_gpu_computing.md

```diff
@@ -0,0 +1,658 @@
+# Distributed GPU Computing
+The scaling of machine learning (ML) workloads in HPC environments can be achieved through various parallelism approaches. This section outlines the primary methods for parallelizing ML computations.
+
+One can decide to use multiple GPUs on their AI/ML applications for different reasons including but not limited to:
+* Handling large-scale datasets
+* Hyperparameter tuning
+* Train/inferece large-scale Model that does not fit into the memory of a single GPU. 
+
+## Inter GPU Communication
+In majority of use-cases of multi-GPU computation there is the need for different GPU to communicate and send their partail computation to one another to sync. NCCL libarary from NVIDIA is widely in use for NIVIDA based GPU communication.
+(sec-nccl)=
+### NVIDIA Collective Communication Library (NCCL)
+For multi-GPU and multi-node communication, NVIDIA Collective Communication Library (NCCL, pronounced “Nickel”) is being used as backend in distributed strategies for Nvidia GPUs such as Distributed Data Parallel (DDP) and Fully Sharded Data Parallel (FSDP). Following are some of the most related NCCL collective communication primitives :
+* Scatter: From one rank, data will be distributed across all rank, with each rank receiving a subpart of the data.
+* Gather: One rank will receive the aggregation of data from all ranks.
+* AllGather: Each rank receives the aggregation of data from all ranks in the order of the ranks.
+* Reduce: One rank receives the reduction of input values across ranks.
+* AllReduce: Each rank receives the reduction of input values across ranks.
+* ReduceScatter: Input values are reduced across ranks, with each rank receiving a subpart of the result.
+```{seealso}
+For more information about the different NCCL collective operations refer to:
+https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html
+``` 
+```{note}
+Each process in the multi-process applications is called a rank. Usually each process has its own exclusive device. Therefore you can think of each rank as one GPU in the following diagrams.
+```
```

Comment:
I see you defined it here, it might be better to move this to before those definition.